src/Evolution/Systems/RelativisticEuler/Valencia/FixConservatives.hpp

```diff
@@ -0,0 +1,127 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <limits>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Evolution/Systems/RelativisticEuler/Valencia/TagsDeclarations.hpp"
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/GeneralRelativity/TagsDeclarations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+class DataVector;
+
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace RelativisticEuler {
+namespace Valencia {
+
+/*!
+ * \brief Fix conservative variables using the method proposed in F. Foucart's
+ * PhD thesis (Cornell)
+ *
+ * Fix the conservative variables according to the following recipe:
+ *
+ * - If \f$D < D_\text{cutoff}\f$, set \f$D = D_\text{min}\f$
+ * - If \f$\tau < 0\f$, set \f$\tau = 0\f$
+ * - If \f$\tilde S^2 > \tilde\tau(\tilde\tau + 2\tilde D)\f$, rescale
+ *   \f$\tilde S_i\f$ by the factor
+ *
+ * \f{align*}
+ * f = \sqrt{\frac{(1 - \epsilon)\tilde S_\text{max}^2}{\tilde S^2 +
+ * 10^{-16}\tilde D^2}} < 1 \f}
+ *
+ * (if \f$ f\geq 1\f$, do not fix \f$\tilde S_i\f$),
+ *
+ * where \f$D_\text{min}\f$, \f$D_\text{cutoff}\f$ and \f$\epsilon\f$
+ * are small and positive input parameters, typically chosen in consistency with
+ * some variable fixing applied to the primitives variables
+ * (e.g. VariableFixing::FixToAtmosphere).
+ */
+template <size_t Dim>
+class FixConservatives {
+ public:
+  /// The minimum value of the rest mass density times the Lorentz factor,
+  /// \f$D\f$
+  struct MinimumValueOfD {
+    using type = double;
+    static constexpr OptionString help = {
+        "Minimum value of rest-mass density times Lorentz factor"};
+    static type lower_bound() noexcept { return 0.0; }
+  };
+
+  /// The cutoff below which \f$D\f$ is set to `MinimumValueOfD`
+  struct CutoffD {
+    using type = double;
+    static constexpr OptionString help = {
+        "Cutoff below which D is set to MinimumValueOfD"};
+    static type lower_bound() noexcept { return 0.0; }
+  };
+
+  /// The safety factor to fix \f$\tilde S_i\f$
+  struct SafetyFactorForS {
+    using type = double;
+    static constexpr OptionString help = {
+        "Safety factor for momentum density bound."};
```

Comment:
This has an upper bound of 1, and lower bound of `epsilon()`. It should also be documented that `1-SafetyFactorForS` is used, since this informs how small is reasonable.